File: python/0074.py
class Solution:
    def solve(self, matrix, x, y, xx, yy, target):
        if target > matrix[xx - 1][yy - 1] or target < matrix[x][y]: return False
        if x + 1 >= xx and y + 1 >= yy:
            if matrix[x][y] == target:
                return True
            else:
                return False
        mx = (x + xx) >> 1
        my = (y + yy) >> 1
        if matrix[mx][my] > target:
            ans = self.solve(matrix, x, y, mx, my, target)
        else:
            ans = self.solve(matrix, mx, my, xx, yy, target)
        if ans: return True
        ans = self.solve(matrix, mx, y, xx, my, target) or self.solve(matrix, x, my, mx, yy, target)
        return ans

    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        if not matrix: return False
        r = len(matrix)
        if not r: return False
        c = len(matrix[0])
        if not c: return False
        return self.solve(matrix, 0, 0, r, c, target)


if __name__ == '__main__':
    print(Solution().searchMatrix([[1, 3]], 1))

File: rust/src/bin/0074.rs
struct Solution;

impl Solution {
    pub fn search_matrix(matrix: Vec<Vec<i32>>, target: i32) -> bool {
        if matrix.len() == 0 {
            return false;
        }
        if matrix[0].len() == 0 {
            return false;
        }
        let mut i = 0;
        let mut j = matrix[0].len() - 1;
        loop {
            if matrix[i][j] < target {
                i += 1;
                if i == matrix.len() {
                    return false;
                }
            } else if matrix[i][j] > target {
                if j == 0 {
                    return false;
                }
                j -= 1;
            } else {
                return true;
            }
        }
    }
}

fn main() {
    let matrix = vec![
        vec![1, 3, 5, 7],
        vec![10, 11, 16, 20],
        vec![23, 30, 34, 50]
    ];
    let ans = Solution::search_matrix(matrix, 17);
    println!("{}", ans);
}